### PR TITLE
New principle: Some APIs should only be exposed to dedicated workers

### LINKED ISSUE #360: Guidance in exposing some APIs only off the main thread
Following up on https://github.com/w3ctag/design-principles/issues/325, which has already quite a bit of background on this:
is it appropriate to use non-exposure of an API on `Window` as a way to force developers to not do things on the main thread?

That question has emerged in [WebCodecs](https://lists.w3.org/Archives/Public/public-media-wg/2021Jul/0017.html), [WebRTC Encoded Transform](https://github.com/w3c/webrtc-encoded-transform/pull/62/files), [Media Capture Transform](https://github.com/w3c/mediacapture-transform/issues/23), [WebNN](https://github.com/webmachinelearning/webnn/issues/229#issuecomment-983751496), [File API](https://w3c.github.io/FileAPI/#dfn-FileReaderSync).

### PR DESCRIPTION
**Pull request by hober**
Fixes #360.

***
<a href="https://pr-preview.s3.amazonaws.com/w3ctag/design-principles/pull/404.html" title="Last updated on Feb 7, 2023, 1:35 AM UTC (f81761f)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3ctag/design-principles/404/e81cae6...f81761f.html" title="Last updated on Feb 7, 2023, 1:35 AM UTC (f81761f)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -2680,21 +2682,59 @@ offers guidance that should be considered in the development of new features, no
...
+
+<h4 id="worker-only">Some APIs should only be exposed to dedicated workers</h4>
+<!--
+     https://github.com/w3c/webcodecs/issues/199#issuecomment-829872273
+     https://github.com/w3c/webcodecs/issues/211
+     https://github.com/w3ctag/design-reviews/issues/612
+     https://github.com/w3ctag/design-principles/issues/325
+     https://github.com/w3ctag/design-principles/issues/360
+-->
+
```
Comment (cynthia) [inline index.bs:2709]: We should have a TL:DR; one-line guidance here. https://github.com/w3ctag/process/blob/master/style-guide.md#get-to-the-point

File: index.bs [hunk 2]
```diff
@@ -2680,21 +2682,59 @@ offers guidance that should be considered in the development of new features, no
 * [[#feature-detect]]
 * Polyfill development should be encouraged
 
-<h3 id="consider-dedicatedworker-support">Where possible APIs should be made available in DedicatedWorker</h3>
+<h3 id="consider-dedicatedworker-support">Where possible APIs should be made available to dedicated workers</h3>
 
 When exposing a feature, please consider whether it makes sense to expose the feature
-to DedicatedWorker as well.
+to dedicated workers (via the {{Worker}} interface).
```
Comment (annevk) [inline index.bs:2688]:
```suggestion
to dedicated workers (via the {{DedicatedWorkerGlobalScope}} interface).
```

File: index.bs [hunk 3]
```diff
@@ -2680,21 +2682,59 @@ offers guidance that should be considered in the development of new features, no
...
+the user experience will suffer.
+(This problem is even more pronounced on low-powered devices,
+which are more likely to be used by disadvantaged or marginalized users.
+Remember,
+[the web is for all people](https://www.w3.org/2001/tag/doc/ethical-web-principles/#allpeople).)
+
+Therefore,
+APIs which would often block the main thread if used as intended
+should not be exposed on the {{Window}} interface.
+By restricting such APIs to the {{Worker}} interface,
```
Comment (annevk) [inline index.bs:2728]:
```suggestion
By restricting such APIs to the {{DedicatedWorkerGlobalScope}} interface,
```

File: index.bs [hunk 4]
```diff
@@ -2680,21 +2682,59 @@ offers guidance that should be considered in the development of new features, no
...
+the “easy” path for web developers
+is the path with the best experience for users. 
+
+<p class="example">
+{{ScriptProcessorNode}}s were replaced by {{AudioWorklet}}s
+in the Web Audio API
+because use of {{ScriptProcessorNode}} from the main thread
+frequently resulted in a poor user experience. [[WebAudio]]
+</p>
```
Comment (annevk) [inline index.bs:2737]: This example suggests worklets are a good alternative, but the text doesn't really go into that.

**Review by annevk — COMMENTED** [review]
It might make sense to focus a bit more on what we don't want (main thread blocking) and then offer a variety of solutions, such as dedicated workers, shared workers, service workers, worklets, and promise-returning APIs.

**Review by cynthia — APPROVED** [review]

**Pull request merged**